File: apps/web/src/domain/repository/healthRepository.ts
/**
 * Health Repository Interface
 *
 * Repository abstraction for health check operations (DIP).
 * This interface is defined in the domain layer and implemented in the adapter layer.
 */

import type { HealthMessage } from "@/domain/model/health";

/**
 * Repository interface for health check operations.
 * Adapter layer must implement this interface.
 */
export interface IHealthRepository {
  /**
   * Save a health check message to the database.
   * @param message - The message to save
   * @returns The saved health message with ID and timestamp
   */
  saveMessage(message: string): Promise<HealthMessage>;

  /**
   * Get the latest health check message from the database.
   * @returns The latest health message, or null if none exists
   */
  getLatest(): Promise<HealthMessage | null>;
}

File: apps/api/src/app/domain/repository/health_repository.py
"""Health Repository Interface.

This module defines the output port for health message persistence.
"""

from abc import ABC, abstractmethod

from domain.model.health_message import HealthMessage


class IHealthRepository(ABC):
    """Repository interface for health messages."""

    @abstractmethod
    async def save(self, message: HealthMessage) -> HealthMessage:
        """Save a health message.

        Args:
            message: The health message to save

        Returns:
            The saved health message

        """

    @abstractmethod
    async def find_latest(self) -> HealthMessage | None:
        """Find the latest health message.

        Returns:
            The latest health message, or None if not found

        """
